shlo_ref: Add dot_general op and unit tests

nishantsarda-mcw opened the pull request:
- Adds initial version of dot_general op
- Adds unit tests for dot_general op (quantized tests are yet to be added)

CC: @rascani @qukhan

nishantsarda-mcw committed:
shlo_ref add `dot_general` op

- Adds initial version of dot_general op
- Adds unit tests for dot_general op

qukhan commented on tensorflow/lite/experimental/shlo/ops/dot_general.cc:
```diff
@@ -0,0 +1,441 @@
+                             const Tensor& lhs_contracting_dimensions,
+                             const Tensor& rhs_contracting_dimensions,
+                             Tensor& output) {
+  const int32_t* lhsb = lhs_batching_dimensions.GetDataAs<DataType::kSI32>();
```
This should be SI64 per the spec. This type is currently missing. A fix is on its way.

nishantsarda-mcw replied:
changed to SI64.

qukhan commented on tensorflow/lite/experimental/shlo/ops/dot_general.cc:
```diff
@@ -0,0 +1,441 @@
+  const size_t lhs_rank = lhs.Rank();
+  const size_t rhs_rank = rhs.Rank();
+  const size_t output_rank = output.Rank();
+  std::vector<size_t> lhs_result_dims;
```
We try to avoid memory allocations. You can use `absl::InlinedVector<T, *REASONABLE_SIZE*>` to have a vector that uses small buffer optimisation.

nishantsarda-mcw replied:
used absl::InlinedVector vectors.

qukhan commented on tensorflow/lite/experimental/shlo/ops/dot_general.cc:
```diff
@@ -0,0 +1,441 @@
+      }
+      auto check_zero_point_value_is_zero =
+          [](auto zero_point) -> bool {
+        if (std::holds_alternative<I4>(zero_point)) {
+          return std::get<I4>(zero_point) == static_cast<I4>(0);
+        } else if (std::holds_alternative<int8_t>(zero_point)) {
+          return std::get<int8_t>(zero_point) == static_cast<int8_t>(0);
+        } else if (std::holds_alternative<int16_t>(zero_point)) {
+          return std::get<int16_t>(zero_point) == static_cast<int16_t>(0);
+        }
+        return false;
```
Does something like this work?
```suggestion
        return std::visit([](const auto v) { return v == static_cast<decltype(v)>(0); }, zero_point);
```

nishantsarda-mcw replied:
this worked.

qukhan commented on tensorflow/lite/experimental/shlo/ops/dot_general.cc:
```diff
@@ -0,0 +1,441 @@
+      auto check_zero_points_values_are_zero =
+          [](auto zero_points) -> bool {
+        if (std::holds_alternative<absl::InlinedVector<I4, 8>>(zero_points)) {
+          auto zero_point_value =
```
`auto&`, otherwise this will make a copy of the vector.

Also add `const` when you don't plan to modify the value.

qukhan commented on tensorflow/lite/experimental/shlo/ops/dot_general.cc:
```diff
@@ -0,0 +1,441 @@
+          }
+        }
+      }
+      auto check_zero_points_values_are_zero =
```
You can also use `std::visit` here to avoid duplicating the code.

nishantsarda-mcw replied:
changed to std::visit

qukhan commented on tensorflow/lite/experimental/shlo/ops/dot_general.cc:
```diff
@@ -0,0 +1,441 @@
+  const size_t output_rank = output.Rank();
+
+  const int32_t* lhs_batching_dimensions_data =
+      lhs_batching_dimensions.GetDataAs<DataType::kSI32>();
```
`SI64`.  This type is currently missing. A fix is on its way.

nishantsarda-mcw replied:
changed to SI64

qukhan commented on tensorflow/lite/experimental/shlo/ops/dot_general.cc:
```diff
@@ -0,0 +1,441 @@
+  DimensionSize lhs_element_index = 0, rhs_element_index = 0;
+  for (DimensionSize k = 0; k < output_size; ++k, ++output_data) {
+    GenerateIndices(k);
+    std::fill(op.lhs_index.begin(), op.lhs_index.end(), 0);
```
`absl::c_fill` can also work and allows you to get rid of the `begin`/`end` pair.

nishantsarda-mcw replied:
changed to absl::c_fill(op.lhs_index, 0);

qukhan commented on tensorflow/lite/experimental/shlo/ops/dot_general.cc:
```diff
@@ -0,0 +1,441 @@
+
+absl::Status Prepare(DotGeneralOp& op, const Tensor& lhs, const Tensor& rhs,
+                     Tensor& output) {
+  if (absl::Status status =
+          CheckParameters(lhs, rhs, op.attributes.lhs_batching_dimensions,
+                          op.attributes.rhs_batching_dimensions,
+                          op.attributes.lhs_contracting_dimensions,
+                          op.attributes.rhs_contracting_dimensions, output);
+      !status.ok()) {
+    return status;
+  }
```
Use `SHLO_REF_RETURN_ON_ERROR`.

https://github.com/tensorflow/tensorflow/blob/b19a54c0d9739f9c548222d625c1413628b3a43f/tensorflow/lite/experimental/shlo/ops/util.h#L26-L29

```suggestion
  SHLO_REF_RETURN_ON_ERROR(CheckParameters(
                          lhs, rhs, op.attributes.lhs_batching_dimensions,
                          op.attributes.rhs_batching_dimensions,
                          op.attributes.lhs_contracting_dimensions,
                          op.attributes.rhs_contracting_dimensions, output));
```

nishantsarda-mcw replied:
used SHLO_REF_RETURN_ON_ERROR.

qukhan commented on tensorflow/lite/experimental/shlo/ops/dot_general.h:
```diff
@@ -0,0 +1,51 @@
+    Tensor rhs_contracting_dimensions;
+  };
+  Attributes attributes;
+  std::vector<size_t> lhs_result_dims;
+  std::vector<size_t> rhs_result_dims;
+  std::vector<size_t> lhs_index;
+  std::vector<size_t> rhs_index;
+  std::vector<size_t> lhs_index_helper;
+  std::vector<size_t> rhs_index_helper;
+  std::vector<size_t> output_index;
+  std::vector<size_t> output_shape;
```
We try to avoid memory allocations.

nishantsarda-mcw replied:
changed to absl::InlinedVector<size_t, 6>

nishantsarda-mcw committed:
Address review comments

rascani commented on tensorflow/lite/experimental/shlo/ops/dot_general.cc:
```diff
@@ -0,0 +1,638 @@
+  auto GenerateIndices = [&](size_t index) -> void {
+    size_t rank = op.output_shape.size();
+    size_t divisor = 1;
+    for (size_t i = 0, j = rank - 1; i < rank; ++i, --j) {
```
This for loop doesn't use `i`, it would be more understandable as `for (size_t j = rank - 1; j >= 0; --j)`

rascani commented on tensorflow/lite/experimental/shlo/ops/dot_general.cc:
```diff
@@ -0,0 +1,638 @@
+    }
+  };
+  // function to incremement lhs and rhs indices
+  auto IncrementIndices = [&]() -> bool {
```
Can we move this to a standalone function?

nishantsarda-mcw replied:
Made this function as a standalone function as suggested.

rascani commented on tensorflow/lite/experimental/shlo/ops/dot_general.cc:
```diff
@@ -0,0 +1,638 @@
+      lhs_element_index = 0;
+      rhs_element_index = 0;
+      for (size_t i = 0; i < lhs_rank; ++i) {
+        lhs_element_index += op.lhs_index[i] * op.lhs_index_helper[i];
```
We should add an accessor to Tensor for addressing an element by a full index. We just didn't have a need for that before with elementwise ops, but it will certainly be the case for more complicated ones like this.

nishantsarda-mcw replied:
Added Tensor::Get and Tensor::Set library methods. now we can use the get and set function to access the data or set the data.

rascani commented on tensorflow/lite/experimental/shlo/ops/dot_general.cc:
```diff
@@ -0,0 +1,638 @@
+}
+
+template <DataType storage_type, DataType expressed_type>
+void DequantizeOpQuantizePerAxis(DotGeneralOp& op, const Tensor& lhs,
```
Similar comment here about dequantizing the whole tensors. When quantized tensors are large, this becomes quite memory expensive.

nishantsarda-mcw replied:
Now this calculations are done in the prepare step and not in the eval.

rascani commented on tensorflow/lite/experimental/shlo/ops/dot_general_test.cc:
```diff
@@ -0,0 +1,1894 @@
+  const Shape shape_rhs({4, 8, 3, 4, 2});
+  const Shape shape_r({8, 4, 3, 2});
+
+  Vector<StorageT> lhs_data{
```
I'd generally suggest moving the test data to a separate file with inline constexpr arrays that can be referenced from the tests. The large initialization blocks become distracting from reading the test bodies themselves.

nishantsarda-mcw replied:
We have made the required changes, will update the same in the next commit.

nishantsarda-mcw committed:
Address review comments
- Addresses the review comments in the dot_general.cc and dot_general_test.cc files
- Adds Tensor::Get and Tensor::Set library methods

nishantsarda-mcw commented on tensorflow/lite/experimental/shlo/ops/dot_general.cc:
```diff
@@ -0,0 +1,638 @@
+    const absl::Span<int64_t> rhs_batching_dimensions,
+    const absl::Span<int64_t> lhs_contracting_dimensions,
+    const absl::Span<int64_t> rhs_contracting_dimensions, Tensor& output,
+    std::array<PrecisionTypes, 2>& precision_configs) {
```
Changed to const.

nishantsarda-mcw commented on tensorflow/lite/experimental/shlo/ops/dot_general.cc:
```diff
@@ -0,0 +1,638 @@
+    return absl::FailedPreconditionError(
+        "stablehlo.dot_general: Size of lhs_batching_dimensions and "
+        "rhs_batching_dimensions must be same.");
+  } else if (lhsc_size != rhsc_size) {
```
removed else as suggested.

nishantsarda-mcw commented on tensorflow/lite/experimental/shlo/ops/dot_general.cc:
```diff
@@ -0,0 +1,638 @@
+  }
+  for (DimensionSize i = 0; i < lhsb_size; ++i) {
+    for (DimensionSize j = 0; j < lhsc_size; ++j) {
+      if (lhs_batching_dimensions[i] == lhs_contracting_dimensions[j]) {
```
Yes the condition was given wrong, now I have changed it.

nishantsarda-mcw commented on tensorflow/lite/experimental/shlo/ops/dot_general.cc:
```diff
@@ -0,0 +1,638 @@
+  }
+  for (DimensionSize i = 0; i < rhsb_size; ++i) {
+    for (DimensionSize j = 0; j < rhsc_size; ++j) {
+      if (rhs_batching_dimensions[i] == rhs_contracting_dimensions[j]) {
```
Updated as above.

nishantsarda-mcw commented on tensorflow/lite/experimental/shlo/ops/dot_general.cc:
```diff
@@ -0,0 +1,638 @@
+        lhs_batching_dimensions[i] < 0) {
+      return absl::FailedPreconditionError(
+          "stablehlo.dot_general: Invalid lhs_batching_dimensions index.");
+    }
```
Used helper function as suggested.

nishantsarda-mcw commented on tensorflow/lite/experimental/shlo/ops/dot_general.cc:
```diff
@@ -0,0 +1,638 @@
+  }
+  for (DimensionSize i = 0; i < lhsb_size; ++i) {
+    if (lhs.shape().Dim(lhs_batching_dimensions[i]) !=
+        rhs.shape().Dim(rhs_batching_dimensions[i])) {
```
Used helper function as suggested.

nishantsarda-mcw commented on tensorflow/lite/experimental/shlo/ops/dot_general.cc:
```diff
@@ -0,0 +1,638 @@
+      return absl::FailedPreconditionError(
+          "stablehlo.dot_general: Invalid lhs_batching_dimensions index.");
+    }
+    output_shape_check.push_back(lhs.shape().Dim(lhs_batching_dimensions[i]));
```
Yes, now the calculations related to expected_output_shape is in one place.

nishantsarda-mcw commented on tensorflow/lite/experimental/shlo/ops/dot_general.cc:
```diff
@@ -0,0 +1,638 @@
+          "contracting dimension size.");
+    }
+  }
+  for (size_t i = 0; i < lhs_rank; ++i) {
```
Used helper function as suggested.

nishantsarda-mcw commented on tensorflow/lite/experimental/shlo/ops/dot_general.cc:
```diff
@@ -0,0 +1,638 @@
+    }
+  }
+  for (size_t i = 0; i < lhs_result_dims.size(); ++i) {
+    output_shape_check.push_back(lhs.shape().Dim(lhs_result_dims[i]));
```
Changed the name to expected_output_shape.

nishantsarda-mcw commented on tensorflow/lite/experimental/shlo/ops/dot_general.cc:
```diff
@@ -0,0 +1,638 @@
+  }
+  if (output_shape_check.size()) {
+    for (size_t i = 0; i < output_rank; ++i) {
+      if (output.shape().Dim(i) != output_shape_check[i]) {
```
Done as suggested.

nishantsarda-mcw commented on tensorflow/lite/experimental/shlo/ops/dot_general.cc:
```diff
@@ -0,0 +1,638 @@
+    output_shape_check.push_back(lhs.shape().Dim(lhs_result_dims[i]));
+  }
+  for (size_t i = 0; i < rhs_result_dims.size(); ++i) {
+    output_shape_check.push_back(rhs.shape().Dim(rhs_result_dims[i]));
```
Used the above to avoid manual loops, which is already done by Shape::Dims(axes)

nishantsarda-mcw commented on tensorflow/lite/experimental/shlo/ops/dot_general.cc:
```diff
@@ -0,0 +1,638 @@
+          "of lhs and rhs must be the same.");
+    }
+  }
+  if (lhs.IsPerTensorQuantized()) {
```
Changed as suggested.

nishantsarda-mcw commented on tensorflow/lite/experimental/shlo/ops/dot_general.cc:
```diff
@@ -0,0 +1,638 @@
+    }
+  }
+  if (lhs.IsPerTensorQuantized()) {
+    if (rhs.IsQuantized() && !output.IsQuantized()) {
```
Changed to the correct constraint as suggested

nishantsarda-mcw commented on tensorflow/lite/experimental/shlo/ops/dot_general.cc:
```diff
@@ -0,0 +1,638 @@
+      return absl::FailedPreconditionError(
+          "stablehlo.dot_general: If the lhs and rhs are quantized tensors, "
+          "than they should have the same storage type.");
+    } else if (rhs.IsPerTensorQuantized()) {
```
Removed the else

nishantsarda-mcw commented on tensorflow/lite/experimental/shlo/ops/dot_general.cc:
```diff
@@ -0,0 +1,638 @@
+              "be the same as the expressed_type of lhs and rhs tensors.");
+        }
+      }
+      auto check_zero_point_value_is_zero = [](const auto& zero_point) -> bool {
```
Made it as a standalone function.

nishantsarda-mcw commented on tensorflow/lite/experimental/shlo/ops/dot_general.cc:
```diff
@@ -0,0 +1,638 @@
+    } else if (rhs.IsPerAxisQuantized()) {
+      for (DimensionSize i = 0; i < rhsc_size; ++i) {
+        if (rhs_contracting_dimensions[i] ==
+            rhs.quantized_per_axis_element_type().QuantizedDimension()) {
```
Used a helper function.

nishantsarda-mcw commented on tensorflow/lite/experimental/shlo/ops/dot_general.cc:
```diff
@@ -0,0 +1,638 @@
+                      op.attributes.precision_configs));
+
+  for (size_t i = 0; i < lhs.Rank(); ++i) {
+    if ((std::count(op.attributes.lhs_batching_dimensions.begin(),
```
Used as suggested

nishantsarda-mcw commented on tensorflow/lite/experimental/shlo/ops/dot_general.cc:
```diff
@@ -0,0 +1,638 @@
+  op.rhs_index_helper.resize(rhs.Rank());
+  op.output_index.resize(output.Rank());
+  for (size_t i = 0; i < output.Rank(); ++i) {
+    op.output_shape.push_back(output.shape().Dim(i));
```
Yes it was redundant, removed it.

nishantsarda-mcw commented on tensorflow/lite/experimental/shlo/ops/dot_general.cc:
```diff
@@ -0,0 +1,638 @@
+  op.rhs_index.resize(rhs.Rank());
+  op.lhs_index_helper.resize(lhs.Rank());
+  op.rhs_index_helper.resize(rhs.Rank());
+  op.output_index.resize(output.Rank());
```
Used InlinedVector as suggested.

srirammaswamy-mcw commented on tensorflow/lite/experimental/shlo/ops/dot_general.h:
```diff
@@ -30,21 +30,21 @@ enum class PrecisionTypes {
+    absl::Span<const Axis> lhs_batching_dimensions;
+    absl::Span<const Axis> rhs_batching_dimensions;
+    absl::Span<const Axis> lhs_contracting_dimensions;
+    absl::Span<const Axis> rhs_contracting_dimensions;
```
@rascani, can we use Axis type (alias of size_t) here, since that will align with the internal functions like `Shape::Dims`. But it will differ from spec.

nishantsarda-mcw commented on tensorflow/lite/experimental/shlo/ops/dot_general.cc:
```diff
@@ -0,0 +1,638 @@
+  const DimensionSize lhsc_size = lhs_contracting_dimensions.size();
+
+  // function to generate indices for output
+  auto GenerateIndices = [&](size_t index) -> void {
```
I have added it as a Tensor::GetNdIndex library method.

nishantsarda-mcw commented on tensorflow/lite/experimental/shlo/ops/dot_general.cc:
```diff
@@ -0,0 +1,638 @@
+}
+
+template <DataType storage_type, DataType expressed_type>
+void DequantizeOpQuantizePerTensor(DotGeneralOp& op, const Tensor& lhs,
```
Now this calculations are done in the prepare step and not in the eval.

srirammaswamy-mcw commented on tensorflow/lite/experimental/shlo/ops/dot_general.cc:
```diff
@@ -26,23 +26,126 @@ limitations under the License.
+  const size_t rhsb_size = rhs_batching_dimensions.size();
+  const size_t lhsc_size = lhs_contracting_dimensions.size();
+  const size_t rhsc_size = rhs_contracting_dimensions.size();
   const size_t lhs_rank = lhs.Rank();
```
@rascani, `Tensor::Rank()` returns `size_t` type, can we change it to its alias type `Axis` so that it will align with other library functions like `Shape::Dim/Shape::Dims` ?

nishantsarda-mcw committed:
Adds Tensor::ExpressedType() method

nishantsarda-mcw commented on tensorflow/lite/experimental/shlo/ops/dot_general.cc:
```diff
@@ -0,0 +1,638 @@
+            "stablehlo.dot_general: If lhs and rhs are per-tensor quantized "
+            "than output should also be per-tensor quantized.");
+      }
+      if (lhs.quantized_per_tensor_element_type().ExpressedType() ==
```
Addressed the above, added Tensor::ExpressedType() library method. 
Addressed in this commit-https://github.com/tensorflow/tensorflow/commit/82b9c104b54d7551cbc8e337140e86a77ea967a0

mihaimaruseac commented:
Closing PRs that have been stale for over a year. Please open again if they still relevant and we will attempt to find faster reviewers.
